### ISSUE TITLE
scale-in and scale-out of node fails due to error in garbage collection of IPs

### ISSUE THREAD

**Issue by hanamantagoudvk** [projectcalico/calico#5287]
We have k8s cluster with calico-cni (v3.20.2) running.

When worker node scale-in is executed and scale-out is performed later , it fails due to below errors.

a) First IP leak is detected for various pod which were located on worker node xyz

```
	 ipam_allocation.go 89: Confirmed IP leak handle="k8s-pod-network.5ff5cd34a3476accbe325a6af65213ddc8c5cc2239a7a13fe36c176023dc9eb1" ip="192.168.72.68" node="xyz"  
	 ipam_allocation.go 89: Confirmed IP leak handle="k8s-pod-network.e58d07fa17e05f267777394886576f640c99f9f7afdc0c0df1aef5b2ac172aec" ip="192.168.72.70" node="xyz"  
	 ipam_allocation.go 89: Confirmed IP leak handle="k8s-pod-network.ef9c4a8af82c3fda187580926bb7e3d81ff4efe90d17900d702dddab67a23a53" ip="192.168.72.71" node="xyz"  
	 ipam_allocation.go 89: Confirmed IP leak handle="k8s-pod-network.9f61664ae4fd001c0902c8b41fd155b5750abeb218d1bdeb53dbb2e90f9af936" ip="192.168.72.74" node="xyz"  
	 ipam_allocation.go 89: Confirmed IP leak handle="k8s-pod-network.2dbae6e94a6713dd4bdcee9a646a3a0a0fc91aabe001e932a851567ece4c9f28" ip="192.168.72.76" node="xyz"  
	 ipam_allocation.go 89: Confirmed IP leak handle="k8s-pod-network.f8b60405afb4e7767ebd4ee4ead51610bdbe1aa861bf96043815f2675f85e8a7" ip="192.168.72.78" node="xyz"  
	 ipam_allocation.go 89: Confirmed IP leak handle="k8s-pod-network.316179400ed38e70490fdd84e1232d0b7916b2c1c8fc6b0fcdd58d95b9b3ef24" ip="192.168.72.65" node="xyz"  
	 ipam_allocation.go 89: Confirmed IP leak handle="ipip-tunnel-addr-xyz" ip="192.168.72.64" node="xyz"
	 ipam_allocation.go 89: Confirmed IP leak handle="ipip-tunnel-addr-xyz" ip="192.168.72.64" node="xyz"
```

b)  Garbage collection is attempted for IPs 

```
	 ipam.go 903: Garbage collecting leaked IP address handle="k8s-pod-network.d517c569061be1810c58f2140cd1b735fde9bffa27e7b25ddcad6e8cc00e8e8a" ip="192.168.72.79" node="xyz"   
	 ipam.go 903: Garbage collecting leaked IP address handle="k8s-pod-network.9f61664ae4fd001c0902c8b41fd155b5750abeb218d1bdeb53dbb2e90f9af936" ip="192.168.72.74" node="xyz"   
         ipam.go 903: Garbage collecting leaked IP address handle="k8s-pod-network.2eab6c82ac744374699791a829926e691d876103d2323b3f733beeabae9a9a9a" ip="192.168.72.66" node="xyz"   
	 ipam.go 903: Garbage collecting leaked IP address handle="k8s-pod-network.5a2ce52e929dbd5204c25334fc654e6adfd160f749988f4bce196838787baee2" ip="192.168.72.75" node="xyz"   
	 ipam.go 903: Garbage collecting leaked IP address handle="k8s-pod-network.316179400ed38e70490fdd84e1232d0b7916b2c1c8fc6b0fcdd58d95b9b3ef24" ip="192.168.72.65" node="xyz"   
	 ipam.go 903: Garbage collecting leaked IP address handle="k8s-pod-network.95af8ef9fc1f3ed4bbcaeff8561059bf467dc1eb9ce34e9b14309e0fc41b48a2" ip="192.168.72.67" node="xyz"   
	 ipam.go 903: Garbage collecting leaked IP address handle="k8s-pod-network.4771eaf97c090a5319d1cf687e8fc2765531e592def18435eef61913eca6c8d0" ip="192.168.72.77" node="xyz"   
	 ipam.go 903: Garbage collecting leaked IP address handle="k8s-pod-network.2dbae6e94a6713dd4bdcee9a646a3a0a0fc91aabe001e932a851567ece4c9f28" ip="192.168.72.76" node="xyz"   
	 ipam.go 903: Garbage collecting leaked IP address handle="k8s-pod-network.9ba91b87d5172e67caaff27dc817bb6d1a2418ecc63a535e1f66a9d1eac8c50e" ip="192.168.72.69" node="xyz" 
```  

c) But the Tunnel IP (ipip-tunnel-addr) has been resurrected.

```
	 ipam.go 897: Leaked IP has been resurrected after querying latest state handle="ipip-tunnel-addr-xyz" ip="192.168.72.64" node="xyz"
	 ipam.go 897: Leaked IP has been resurrected after querying latest state handle="ipip-tunnel-addr-xyz" ip="192.168.72.64" node="xyz"
```

d) Due to stale tunnel-ip , IP block is not empty and hence cleaning up of node fails

```
	 ipam_block_reader_writer.go 300: Block must be empty but is not empty, refusing to remove affinity. host="xyz" subnet="192.168.72.64/26"
	 ipam.go 931: Failed to release block affinities for node calicoNode="xyz" error=block '192.168.72.64/26' is not empty
	 ipam.go 875: Error cleaning up node error=block '192.168.72.64/26' is not empty node="xyz"
```

## Expected Behavior

When node is scaled-in , all IPs including tunnel-ip must be released and calico node must be cleaned up.

## Your Environment

* Calico version : v3.20.2
* Orchestrator version (e.g. kubernetes, mesos, rkt):  k8s 1.22

**Comment by caseydavenport** [comment]
@hanamantagoudvk thanks for the issue - are you seeing that the IP is leaked? Or just that the error is reported?

I think what is happening is that once the new node starts, the IP for the node is valid again and thus so is the block, so the garbage collector is correctly not releasing affinity for the block.

What sort of timescale is this happening on? i.e., how quickly are you adding / deleting this node?

**Comment by hanamantagoudvk** [comment]
@caseydavenport , how do we see if tunnel-ip alone is leaked after node is scaled-in ? What is the cmd ?
Let me know what all info needs to be collected ,so that i can get those info from my test team.

Regarding timescale, i will check with my test folks what is time gap between scale-in and scale-out.

**Comment by caseydavenport** [comment]
There's a calicoct command that does a full IPAM database consistency check you could use: https://projectcalico.docs.tigera.io/reference/calicoctl/ipam/check

Alternatively, you should be able to check the output of `kubectl get ipamhandles -o yaml ipip-tunnel-addr-xyz` to look for this field: https://github.com/projectcalico/calico/blob/master/libcalico-go/lib/apis/v3/ipam_handle.go#L43

The `block` map is a map of block -> # of addresses allocated with that handle in that block. You should only ever see one address per handle if all is well. 

Further information on the allocation can be found with `kubectl get ipamblocks` and `kubectl get blockaffinity` - the `ipamblock` contains the actual allocation state (although is a bit hard to parse) and the `blockaffinity` is a mapping of a block to a particular node.

**Comment by hanamantagoudvk** [comment]
@caseydavenport , the important info i got about this ticket is that, this issue is seen when calico uses etcd as its datastore.

**Comment by shcherbak** [comment]
the same issue on etcd 3.5.2, k8s 1.22.6, calico 3.22.1, debian 10, kernel 5.10

**Comment by shcherbak** [comment]
[hanamantagoudvk](https://github.com/hanamantagoudvk) can you please provide you maxPods value of kubelet and calico CALICO_IPV4POOL_BLOCK_SIZE ?
